---
title: "Differential Expression"
author: "Timothy Smyth"
date: "2025-06-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Clear the environment and import required packages

```{r}
rm(list = ls(all.names = TRUE)) # clears global environ.

library(Biobase)
library(variancePartition)
library(BiocParallel)
library(openxlsx)
library(tidyverse)
library(EnhancedVolcano)
library(eulerr)
```

### Create functions for future visualization

```{r}
plot_volcano <- function(DEG_Input, Comparison_names, Top_Gene_Number, xlim, point_size, text_size, file_name){
  
  # Determine the number of significantly upregulated genes
  up <- DEG_Input %>% count(log2FC >= 1.5 & adj.P.Val < 0.1)
  
  up <- as.numeric(up[2, 2])
  
  # Determine the number of significantly downregulated genes
  down <- DEG_Input %>% count(log2FC <= -1.5 & adj.P.Val < 0.1)
  
  down <- as.numeric(down[2, 2])
  
  # Arrange rows by descending log2FC value
  DEG_Input <- DEG_Input %>% arrange(desc(log2FC))
  
  # Replace empty HUGO IDs with Ensembl IDs
  # DEG_Input$GENE[DEG_Input$GENE == ''] = DEG_Input$GENEID[DEG_Input$GENE == '']
  
  # Isolate upregulated gene names
  selectLab_up <- DEG_Input %>% 
    subset(log2FC >= 1.5 & adj.P.Val < 0.1) %>% 
    arrange(desc(log2FC)) %>% 
    rownames()
  
  # Retain top upregulated genes
  selectLab_up <- selectLab_up[1:Top_Gene_Number] %>% na.omit()
  
  # Isolate downregulated gene names
  selectLab_down <- DEG_Input %>% 
    subset(log2FC <= -1.5 & adj.P.Val < 0.1) %>% 
    arrange(log2FC) %>% 
    rownames()
  
  # Retain top downregulated genes
  selectLab_down <- selectLab_down[1:Top_Gene_Number] %>% na.omit()
  
  selectLabs <- c(selectLab_up, selectLab_down)
  
  DEG_Input[!rownames(DEG_Input) %in% selectLabs, 'GENE'] <- ''
  
  # Remove labels for genes from X/Y chromosomes
  # DEG_Input[which(DEG_Input$CHR == 'X' | DEG_Input$CHR == 'Y'), 'GENE'] <- ''
  
  volcano <- EnhancedVolcano(DEG_Input,
                             lab = DEG_Input$GENE,
                             labSize = text_size,
                             drawConnectors = TRUE,
                             widthConnectors = 0.5,
                             max.overlaps = Inf,
                             maxoverlapsConnectors = Inf,
                             boxedLabels = FALSE,
                             x = 'log2FC',
                             y = 'adj.P.Val',
                             FCcutoff = 1.5,
                             pCutoff = 0.1,
                             title = Comparison_names,
                             pointSize = point_size) 
  
  # Determine the y axis limit
  max <- max(volcano[["plot_env"]][["ylim"]]) * 1.1
  
  # Add arrows and numbers describing the number of up or downregulated genes
  volcano <- volcano + 
    
    # Set x axis from -10 to 10 and extend y axis
    ggplot2::coord_cartesian(xlim = c(-xlim, xlim), 
                             ylim = c(0, max)) +
    
    # Place downregulated numbers on plot
    ggplot2::geom_text(x = -xlim * 0.75,
                       y = max* 0.95,
                       label = paste0(down),
                       size = 8.5) +
    
    # Place downregulation arrow on plot
    ggplot2::geom_segment(x = -xlim * 0.95,
                          y = max,
                          xend = -xlim * 0.95,
                          yend = max * 0.9,
                          arrow = arrow(length = unit(2, "mm")), 
                          linewidth = 1) +
    
    # Place upregulated numbers on plot
    ggplot2::geom_text(x = xlim * 0.75,
                       y = max * 0.95,
                       label = paste0(up),
                       size = 8.5) + 
    
    # Place upregulation arrow on plot
    ggplot2::geom_segment(x = xlim * 0.6,
                          y = max * 0.9,
                          xend = xlim * 0.6,
                          yend = max,
                          arrow = arrow(length = unit(2, "mm")), 
                          linewidth = 1)
  
  # Save the resulting plot
  tiff(paste0(file_name, '.tiff'),
       units = "cm",
       width = 18,
       height = 16,
       res = 300)

  print(volcano)

  dev.off()
  
}

###########################################################################################
###########################################################################################
###########################################################################################

make_euler <- function(Input_list, Plot_name){
  
  # Will isolate upregulated and downregulated genes for separate plots
  list_direction <- list('Up', 'Down')
  
  # Create euler plots for up or downregulated genes
  lapply(list_direction, function(y){
    
    # Create list containing unique gene names from each group
    # This list has only up or downregulated genes at a given time
    eul_list <- list()
    
    for (i in 1:length(Input_list)){
      
      eul_list[[i]] <- unique(Input_list[[i]][[y]])
      
    }
    
    
    # Set names of list to match source group
    names(eul_list) <- names(Input_list)
    
    # Define colors for euler plots
    colors <- c('blue3',
                'khaki3',
                'firebrick3',
                'purple3',
                'seagreen3',
                'darkorange3', 
                
                'antiquewhite3', 
                'cornflowerblue', 
                'darkslategray3', 
                'lightpink2')
    
    # Set group colors
    col <- colors[1:length(eul_list)]
    
    # Fit the euler plot
    fit <- euler(eul_list)
    
    # Add fills, legend, text size, axis limits
    fit <- plot(fit,
                fills = list(fill = col, alpha = 0.5),
                legend = list(side = 'bottom', col = col, text.col = 'black'),
                quantities = list(cex = 1), 
                # lty = 2:1,
                edges = 1,
                # labels = list(font = 4),
                main = paste0(y, 'regulated Genes '),
                xlim = c(-15, 15), 
                ylim = c(-15, 15))
    
    # Determine Figure name
    title <- ifelse(y == 'Up', 
                    Plot_name[1], 
                    Plot_name[2])
    
    # Export as a tiff
    tiff(paste0(title, '.tiff'),
         units = "in",
         width = 7,
         height = 5.5,
         res = 300)

    print(fit)

    dev.off()
    
    ##################################################################################################
    # Determine the genes present within each section of overlap and output gene lists as a csv file #
    ##################################################################################################
    
    # Defining overlap of genes can be easily done using ggVennDiagram functions
    Overlap_venn <- ggVennDiagram:::Venn(eul_list) %>% ggVennDiagram::process_data()

    # Isolate genes unique to each group and combine to one data frame, removing NA values
    # generated from combining unequal length vectors into one data frame
    Gene_overlap <- Overlap_venn[["regionLabel"]][["item"]]
    Gene_overlap <- do.call(qpcR:::cbind.na, Gene_overlap) %>% data.frame()
    Gene_overlap[is.na(Gene_overlap)] <- ''

    # Rename to define overlapping groups
    names(Gene_overlap) <- Overlap_venn[["regionLabel"]][["name"]]

    # write.csv(Gene_overlap, file = paste0(title, '_Gene_Overlap.csv'))
    
    })
}
```

### Import and format metadata

```{r}
set.seed(12345)

# Import metadata
Metadata <- read.table("sample_info_full.txt", 
                       sep = "\t", 
                       header = T)

# Set rownames to Sample ID
rownames(Metadata) <- Metadata$SampleID

# Set donor, timepoint, exposure, and infection to leveled factors
Metadata$donor <- factor(Metadata$donor, 
                         levels = c("1", "2", "3", "4", "5", "6", "7"))

Metadata$timepoint <- factor(Metadata$timepoint, 
                             levels = c("2", "24"))

Metadata$sex <- factor(Metadata$sex, 
                       levels = c("male", "female"))

Metadata$exposure <- factor(Metadata$exposure, 
                            levels = c("control", "winter", "summer", "spring", "fall"))

Metadata$infection <- factor(Metadata$infection, 
                             levels = c("mock", "virus"))

# Creating categories for subgroup analysis
Metadata$Group1 <- paste(Metadata$exposure, 
                         Metadata$infection, 
                         Metadata$timepoint, 
                         sep = "_")

Metadata$Group <- paste(Metadata$sex, 
                         Metadata$exposure, 
                         Metadata$infection, 
                         Metadata$timepoint, 
                         sep = "_")
```

### Import human gene count data

```{r}
# Import count data
count <- read.table("human_geneCount.txt", 
                    sep = "\t", 
                    header = T, 
                    stringsAsFactors = F, 
                    check.names = F, 
                    quote = "")

# Set rownames to ENSEMBL gene ID
rownames(count) <- count$GENEID

# Isolate gene metadata
gene <- count[c(1:8)]

# Remove single quote from gene description
gene$DESCRIPTION <- gsub("\"",
                         "", 
                         gene$DESCRIPTION)

# Remove gene metadata
count <- as.matrix(count[-c(1:8)])

# Rearrange count data to match metadata Sample ID order
count <- count[, Metadata$SampleID]
```

### Import human TPM data

```{r}
# Load transcript per million (TPM) data
tpm <- read.table("human_gene_tpm.txt", 
                  sep = "\t",
                  header = T,
                  stringsAsFactors = F, 
                  check.names = F, 
                  quote = "")

# Set rownames to gene ID
rownames(tpm) <- tpm$GENEID

# Match row ordering to gene metadata
tpm <- tpm[match(gene$GENEID, tpm$GENEID),]

# Remove metadata
tpm <- as.matrix(tpm[-c(1:8)])

# Rearrange columns to match metadata Sample ID order
tpm <- tpm[, Metadata$SampleID]

########################################################################################
### Preprocessing: quantile normalization, log transformation, filtering by mean TPM ###
########################################################################################

# Calculate gene wise counts
sum <- rowSums(count)

# Retain count data with minimal gene wise counts above 100
count <- count[sum > 100, ]

# Retain gene metadata with minimal gene wise counts above 100
gene <- gene[sum > 100, ]

# Retain tpm data with minimal gene wise counts above 100
tpm <- tpm[sum > 100, ]
```

### Calculate group means and create metadata and gene objects for differential expression

```{r}
group <- levels(as.factor(Metadata$Group))

# Create a matrix with dimmensions equal to number of genes and number of groups
m <- matrix(NA, 
             nrow = nrow(tpm), 
             ncol = length(group))

for (i in 1:length(group)) {
  
  # Isolate tpm data for a single group and calculate means for each gene
  m[, i] <- rowMeans(tpm[, Metadata$Group == group[i]])
  
}

# Set colnames to sample group
colnames(m) <- paste(group, 
                      "mean", 
                      sep = "_")

# Set rownames to gene name
rownames(m) <- rownames(tpm)

# Add metadata to mean counts
ann <- cbind(gene, as.data.frame(m))

# Construct gene expression object and do differential expression analysis
MetadataData <- new("AnnotatedDataFrame", data = Metadata)
geneAnnotation <- new("AnnotatedDataFrame", data = ann)

# Create expression set object to be input for LIMMA analysis.
eset <- ExpressionSet(assayData = count, MetadataenoData = MetadataData, featureData = geneAnnotation)

# Define parameter for parallel processing
param = SnowParam(10, "SOCK", progressbar = TRUE, exportglobals = FALSE)
```

### Perform differenital expression analysis of indicated contrasts

```{r}
# Create model for dream
form <- ~0 + Group1 + (1|donor)

# Create design matrix to spot check model
design <- model.matrix(~0 + Group1, data = Metadata)

# Normalize data by voom with dream
group_vm <- voomWithDreamWeights(eset, form, Metadata, BPPARAM = param)

# Create contrast matrix for dream
contrasts <- makeContrastsDream(form, 
                                Metadata,
                                contrasts = c(
                                  
                                  ################ Viral Infection ###############

                                  Influenza_2h_Post_Infection = "(Group1control_virus_2 - Group1control_mock_2)",
                                  
                                  Influenza_24h_Post_Infection = "(Group1control_virus_24 - Group1control_mock_24)",
                                  
                                  Influenza_24h_vs_2h_Post_Infection = "(Group1control_virus_24 - Group1control_mock_24) -
                                                                    (Group1control_virus_2 - Group1control_mock_2)",

                                  ################ Seasonal PM ###############
                                  
                                  Spring_PM.2h = "(Group1spring_mock_2 - Group1control_mock_2)",
                                  Summer_PM.2h = "(Group1summer_mock_2 - Group1control_mock_2)",
                                  Fall_PM.2h = "(Group1fall_mock_2 - Group1control_mock_2)",
                                  Winter_PM.2h = "(Group1winter_mock_2 - Group1control_mock_2)",
                                  
                                  Spring_PM.24h = "(Group1spring_mock_24 - Group1control_mock_24)",
                                  Summer_PM.24h = "(Group1summer_mock_24  - Group1control_mock_24)",
                                  Fall_PM.24h = "(Group1fall_mock_24 - Group1control_mock_24)",
                                  Winter_PM.24h = "(Group1winter_mock_24 - Group1control_mock_24)",

                                  ################ PM v Ctr x Infection ################ 
                                  
                                  Spring_PM_Virus_vs_Vehicle_Virus.2h = "(Group1spring_virus_2 - Group1control_virus_2)",
                                  Summer_PM_Virus_vs_Vehicle_Virus.2h = "(Group1summer_virus_2 - Group1control_virus_2)",
                                  Fall_PM_Virus_vs_Vehicle_Virus.2h = "(Group1fall_virus_2 - Group1control_virus_2)",
                                  Winter_PM_Virus_vs_Vehicle_Virus.2h = "(Group1winter_virus_2 - Group1control_virus_2)",
                                  
                                  Spring_PM_Virus_vs_Vehicle_Virus.24h = "(Group1spring_virus_24 - Group1control_virus_24)",
                                  Summer_PM_Virus_vs_Vehicle_Virus.24h = "(Group1summer_virus_24 - Group1control_virus_24)",
                                  Fall_PM_Virus_vs_Vehicle_Virus.24h = "(Group1fall_virus_24 - Group1control_virus_24)",
                                  Winter_PM_Virus_vs_Vehicle_Virus.24h = "(Group1winter_virus_24 - Group1control_virus_24)"))

# Fit model with dream weights
group_fitmm <- dream(group_vm, form, Metadata, contrasts, ddf = "Satterthwaite", BPPARAM = param)

# Empirical Bayes Statistics for Differential Expression
group_fitmm <- eBayes(group_fitmm)

##########################################################

DEG_contrasts <- colnames(contrasts)

# Calculate DEGs
DEGs <- lapply(DEG_contrasts, function(x){
  
  topTable(group_fitmm, coef = x, sort.by = "P", n = Inf)
  
})

# Results are the following information:
# logFC: log2 fold change of group1/group0
# AveExpr: Average expression across all samples, in log0 CPM
# t: logFC divided by its standard error
# P.Value: Raw p-value (based on t) from test that logFC differs from 0
# adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value
# B: log-odds that gene is DE

# Isolate key DEG readouts
DEG_Results <- lapply(DEGs, function(x){
  
  tmp <- data.frame(cbind(x[["logFC"]], 
                          x[["P.Value"]],
                          x[["adj.P.Val"]],
                          x["AveExpr"]),
                    row.names = rownames(x)) 
  
  colnames(tmp) <- c('log2FC', 'P.Value', 'adj.P.Val', 'Expression')
  
  tmp <- cbind(gene[rownames(tmp), c('GENEID', 'GENE', 'GENETYPE', 'DESCRIPTION')], tmp)

})

names(DEG_Results) <- DEG_contrasts %>% 
  gsub(pattern = '\\.', replacement = ', ') %>% 
  gsub(pattern = 'Influenza_', replacement = '') %>% 
  gsub(pattern = '_Virus', replacement = ' + Influenza') %>% 
  gsub(pattern = '_', replacement = ' ')

export_file_name <- c('Figure_1A', 'Figure_1B', 'Figure_1C', 
                      'Figure_3A', 'Figure_3B', 'Figure_3C', 'Figure_3D', 
                      'Supplemental_Figure_3A', 'Supplemental_Figure_3B', 'Supplemental_Figure_3C', 'Supplemental_Figure_3D', 
                      'Figure_4A', 'Figure_4B', 'Figure_4C', 'Figure_4D', 
                      'Supplemental_Figure_5A', 'Supplemental_Figure_5B', 'Supplemental_Figure_5C', 'Supplemental_Figure_5D')

save(DEG_Results, file = 'DEG_Results.RData')
```

### Create volcano plots presented in Figures 1, 3, and 4 as well as Supplemental Figure S3 and S5

```{r}
for(i in 1:length(DEG_Results)){
  
  graph_title <- names(DEG_Results)[[i]] %>% 
    gsub(pattern = 'PM, 2h', replacement = 'PM') %>% 
    gsub(pattern = 'PM, 24h', replacement = 'PM') %>% 
    gsub(pattern = ' vs Vehicle \\+ Influenza, 2h', replacement = '') %>% 
    gsub(pattern = ' vs Vehicle \\+ Influenza, 24h', replacement = '')

  plot_volcano(DEG_Input = DEG_Results[[i]],
               Comparison_names = graph_title,
               Top_Gene_Number = 20,
               xlim = 10,
               point_size = 3,
               text_size = 2.5, 
               file_name = export_file_name[[i]])

}
```

### Isolate up and downregulated DEGs

```{r}
# Isolate up and downregulated gene names for each tested group
Comparisons <- lapply(names(DEG_Results), function(x){
  
  # Isolate current data set
  tmp_data <- DEG_Results[[x]]
  
  # Replace empty HUGO IDs with Ensembl IDs
  tmp_data$GENE[tmp_data$GENE == ''] = tmp_data$GENEID[tmp_data$GENE == '']
  
  # Isolate up and downregulated genes
  up <- tmp_data %>% subset(adj.P.Val < 0.1 & (log2FC >= 1.5))
  up <- up$GENE
  
  down <- tmp_data %>% subset(adj.P.Val < 0.1 & (log2FC <= -1.5))
  down <- down$GENE
  
  tmp <- list(up, down)
  names(tmp) <- c('Up', 'Down')
  tmp
  
})

names(Comparisons) <- names(DEG_Results)
```

### Create Euler plots for indicated figure pannels

```{r}
############# Viral Exposure Plots ############# 

Viral <- list(Comparisons[['2h Post Infection']], 
              Comparisons[['24h Post Infection']], 
              Comparisons[['24h vs 2h Post Infection']])

names(Viral) <- c('2h Post Infection', 
                  '24h Post Infection', 
                  '24h vs 2h Post Infection')

Viral <- make_euler(Input_list = Viral, 
                    Plot_name = c('Figure_1D', 'Figure_1E'))

############# PM Exposure Plots ############# 

Exposure.2 <- list(Comparisons[["Spring PM, 2h"]], 
                   Comparisons[["Summer PM, 2h"]], 
                   Comparisons[["Fall PM, 2h"]], 
                   Comparisons[["Winter PM, 2h"]])

names(Exposure.2) <- c('Spring PM', 
                       'Summer PM', 
                       'Fall PM', 
                       'Winter PM')

Exposure.2 <- make_euler(Input_list = Exposure.2, 
                         Plot_name = c('Figure_3E', 'Figure_3F'))

####

Exposure.24 <- list(Comparisons[["Spring PM, 24h"]], 
                    Comparisons[["Summer PM, 24h"]], 
                    Comparisons[["Fall PM, 24h"]], 
                    Comparisons[["Winter PM, 24h"]])

names(Exposure.24) <- c('Spring PM', 
                        'Summer PM', 
                        'Fall PM', 
                        'Winter PM')

Exposure.24 <- make_euler(Input_list = Exposure.24, 
                          Plot_name = c('Supplemental_Figure_3E', 'Supplemental_Figure_3F'))

############# PM v Ctr x Infection ############# 

Viral.PM.Control.Virus.2 <- list(Comparisons[['Spring PM + Influenza vs Vehicle + Influenza, 2h']], 
                                 Comparisons[['Summer PM + Influenza vs Vehicle + Influenza, 2h']], 
                                 Comparisons[['Fall PM + Influenza vs Vehicle + Influenza, 2h']], 
                                 Comparisons[['Winter PM + Influenza vs Vehicle + Influenza, 2h']])

names(Viral.PM.Control.Virus.2) <- c('Spring PM + Influenza', 
                                     'Summer PM + Influenza', 
                                     'Fall PM + Influenza', 
                                     'Winter PM + Influenza')

Viral.PM.Control.Virus.2 <- make_euler(Input_list = Viral.PM.Control.Virus.2, 
                                       Plot_name = c('Figure_4E', 'Figure_4F'))

############# PM v Virus ############# 

PM.Vs.PM.Virus.2 <- list(Comparisons[['Spring PM + Influenza vs Vehicle + Influenza, 2h']], 
                         Comparisons[['Summer PM + Influenza vs Vehicle + Influenza, 2h']], 
                         Comparisons[['Fall PM + Influenza vs Vehicle + Influenza, 2h']], 
                         Comparisons[['Winter PM + Influenza vs Vehicle + Influenza, 2h']], 
                         
                         Comparisons[["Winter PM, 2h"]])

names(PM.Vs.PM.Virus.2) <- c('Spring PM + Influenza, 2h', 
                             'Summer PM + Influenza, 2h', 
                             'Fall PM + Influenza, 2h', 
                             'Winter PM + Influenza, 2h',
                             'Winter PM, 2h')

PM.Vs.PM.Virus.2 <- make_euler(Input_list = PM.Vs.PM.Virus.2, 
                               Plot_name = c('Supplemental_Figure_4A', 'Supplemental_Figure_4B'))

############# Virus vs PM and Virus ############# 

Virus.Vs.PM.Virus.2 <- list(Comparisons[['Spring PM + Influenza vs Vehicle + Influenza, 2h']], 
                            Comparisons[['Summer PM + Influenza vs Vehicle + Influenza, 2h']],
                            Comparisons[['Fall PM + Influenza vs Vehicle + Influenza, 2h']], 
                            Comparisons[['Winter PM + Influenza vs Vehicle + Influenza, 2h']], 
                            
                            Comparisons[['2h Post Infection']])

names(Virus.Vs.PM.Virus.2) <- c('Spring PM + Influenza, 2h', 
                                'Summer PM + Influenza, 2h', 
                                'Fall PM + Influenza, 2h', 
                                'Winter PM + Influenza, 2h',
                                
                                'Influenza, 2h')

Virus.Vs.PM.Virus.2 <- make_euler(Input_list = Virus.Vs.PM.Virus.2, 
                                  Plot_name = c('Supplemental_Figure_4C', 'Supplemental_Figure_4D'))
```